### PR TITLE
Delete old rulesets when encountering quota errors

### PR DESCRIPTION
**Pull request by ryanpbrewster**
If we encounter a quota error while uploading rules, we now:
  - count the number of rulesets (this will always be 2501, because that's what causes the quota errors)
  - if there are a lot (> 1000), delete some old rulesets (10) and retry the upload

The happy path is unchanged.

# Happy path
happy

# Current quota behavior
old-sad

# New quota behavior
new-sad

### CODE REVIEW

**Comment by ryanpbrewster** [comment]
I believe this PR will require https://github.com/firebase/firebase-tools/pull/1189 to go in first

File: src/gcp/rules.ts [hunk 1]
```diff
@@ -97,6 +97,20 @@ export async function listRulesets(
   return _handleErrorResponse(response);
 }
 
+/**
+ * Lists all the rulesets for the given project. May require many network requests.
+ */
+export async function listAllRulesets(projectId: string): Promise<object[]> {
+  let pageToken = undefined;
+  let rulesets: object[] = [];
+  do {
```
Comment (mbleigh) [inline src/gcp/rules.ts]: I literally can't think of another time I've seen a `do { } while ()` in JS code...I don't necessarily have a problem with it but it makes me slightly uneasy :grimacing:

File: src/RulesDeploy.js [hunk 2]
```diff
@@ -65,7 +74,30 @@ RulesDeploy.prototype = {
...
+        const history = await gcp.rules.listAllRulesets(self.options.project);
+        if (history.length > RULESET_COUNT_LIMIT) {
+          clc.yellow(
+            `too many rulesets (${history.length}), deleting some old ones to free up space...`
+          );
+          utils.logBullet(
+            clc.bold.yellow(self.type + ":") +
+              ` deleting ${RULESETS_TO_GC} oldest rules (of ${history.length})`
+          );
+          for (let entry of history.reverse().slice(0, RULESETS_TO_GC)) {
```
[the remark this answers is not in the document]
Comment (ryanpbrewster) [reply src/RulesDeploy.js]: It doesn't seem like it's part of our explicit API contract. I'm happy to document and enforce that on the backend, but it's easy and cheap to double-check this on our end. I'll go ahead and `_.sort` it.

File: src/RulesDeploy.js [hunk 3]
```diff
@@ -65,7 +74,30 @@ RulesDeploy.prototype = {
...
+            const rulesetId = entry.name.split("/").pop();
+            await gcp.rules.deleteRuleset(self.options.project, rulesetId);
```
[the remark this answers is not in the document]
Comment (ryanpbrewster) [reply src/RulesDeploy.js]: I was a bit worried about rate limiting. The backend limits creates/deletes to ~1.2 QPS.

I initially considered deleting N rulesets (where N is the number of rules the user is trying to upload), but thought I'd give myself a bit of headroom.

File: src/gcp/rules.ts (hunk 1 above)
Comment (ryanpbrewster) [reply src/gcp/rules.ts]: I basically only ever use do/while loops when doing pagination.

File: src/RulesDeploy.js [hunk 4]
```diff
@@ -65,7 +74,30 @@ RulesDeploy.prototype = {
...
-    return Promise.all(promises);
+    return Promise.all(promises).catch(async (err) => {
+      if (err.status === QUOTA_EXCEEDED_STATUS_CODE) {
+        utils.logBullet(
+          clc.bold.yellow(self.type + ":") + " quota exceeded error while uploading rules"
+        );
+        const history = await gcp.rules.listAllRulesets(self.options.project);
+        if (history.length > RULESET_COUNT_LIMIT) {
+          clc.yellow(
+            `too many rulesets (${history.length}), deleting some old ones to free up space...`
```
[the remark this answers is not in the document]
Comment (bkendall) [reply src/RulesDeploy.js]: see `src/prompt.js` (IIRC)

File: src/RulesDeploy.js [hunk 5]
```diff
@@ -65,7 +74,30 @@ RulesDeploy.prototype = {
         })
       );
     });
-    return Promise.all(promises);
+    return Promise.all(promises).catch(async (err) => {
+      if (err.status === QUOTA_EXCEEDED_STATUS_CODE) {
+        utils.logBullet(
+          clc.bold.yellow(self.type + ":") + " quota exceeded error while uploading rules"
```
Comment (bkendall) [inline src/RulesDeploy.js]: if the RulesDeploy to TS PR goes in first, remove references to `self` (which can already be done after using arrow functions, but definitely should be updated when moving to TS)

File: src/RulesDeploy.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (ryanpbrewster) [reply src/RulesDeploy.js]: Resolving this for now, I expect this codepath to be executed rarely, and deleting 10 rulesets is pretty fast (on the order of a few seconds). I don't want to add retry + backoff logic to circumvent rate limiting. If you feel strongly about this, re-open and I'll make it concurrent.

File: src/RulesDeploy.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (ryanpbrewster) [reply src/RulesDeploy.js]: Right now this code always asks before deleting rulesets. We currently use `--force` for functions, and I'm not sure how I feel about piggy-backing on that, and I'm not eager to introduce a separate flag.

I have a minor preference to merge this as-is, and potentially implement non-interactive mode in a follow-up PR. This PR has important and fairly subtle logic around the Rules API, and I'd like to focus on that.

File: src/RulesDeploy.ts [hunk 6]
```diff
@@ -66,7 +77,39 @@ export class RulesDeploy {
...
+    return Promise.all(promises).catch(async (err) => {
+      if (err.status === QUOTA_EXCEEDED_STATUS_CODE) {
+        utils.logBullet(
+          clc.bold.yellow(this.type + ":") + " quota exceeded error while uploading rules"
+        );
+
+        const history: ListRulesetsEntry[] = await gcp.rules.listAllRulesets(this.options.project);
+
+        if (history.length > RULESET_COUNT_LIMIT) {
+          const shouldContinue = await prompt.once({
```
Comment (mbleigh) [inline src/RulesDeploy.ts]: This should be short-circuited if `--force` flag has been passed.

File: src/RulesDeploy.ts [hunk 7]
```diff
@@ -66,7 +77,39 @@ export class RulesDeploy {
...
+            message: `You have ${
+              history.length
+            } rules, do you want to delete the oldest ${RULESETS_TO_GC} to free up space?`,
+            default: false,
+          });
+          if (shouldContinue) {
+            // Find the oldest unreleased rulesets. The rulesets are sorted reverse-chronlogically.
+            const releases: Release[] = await gcp.rules.listAllReleases(this.options.project);
+            const isReleased = (ruleset: ListRulesetsEntry) =>
+              !!_.find(releases, (release) => release.rulesetName === ruleset.name);
```
Comment (mbleigh) [inline src/RulesDeploy.ts]: Don't have to use lodash here, `releases.find(release => release.rulesetName === ruleset.name)` is fine.

File: src/RulesDeploy.ts [hunk 8]
```diff
@@ -66,7 +77,39 @@ export class RulesDeploy {
...
+            const unreleased: ListRulesetsEntry[] = _.reject(history, isReleased);
+            const entriesToDelete = unreleased.reverse().slice(0, RULESETS_TO_GC);
+            for (const entry of entriesToDelete) {
+              await gcp.rules.deleteRuleset(this.options.project, gcp.rules.getRulesetId(entry));
```
Comment (mbleigh) [inline src/RulesDeploy.ts]: Maybe add 

```
logger.debug(`[rules] Deleted ruleset ${gcp.rules.getRulesetId(entry)}`)
```

Here so the debug logs will show a bit more progress while waiting for these.

File: src/RulesDeploy.ts (hunk 7 above)
Comment (ryanpbrewster) [reply src/RulesDeploy.ts]: done

File: src/RulesDeploy.ts (hunk 8 above)
Comment (ryanpbrewster) [reply src/RulesDeploy.ts]: done

File: src/RulesDeploy.ts (hunk 6 above)
Comment (ryanpbrewster) [reply src/RulesDeploy.ts]: done, not sure exactly how `prompt` is _supposed_ to be used, but this works

File: src/gcp/rules.ts [hunk 9]
```diff
@@ -27,32 +27,69 @@ export async function getLatestRulesetName(
   projectId: string,
   service: string
 ): Promise<string | null> {
+  const releases = await listAllReleases(projectId);
+  const prefix = "projects/" + projectId + "/releases/" + service;
```
Comment (bkendall) [inline src/gcp/rules.ts]: Prefer template strings.

```suggestion
  const prefix = `projects/${projectId}/releases/${service}`;
```

Comment (ryanpbrewster) [reply src/gcp/rules.ts]: done

**Comment by bakerthomas317** [comment]
Theses two cell phone number 5197068848 and 5193851008 are hacking the shit
out of my cell phone can you please help me out

**Pull request merged**